### ISSUE TITLE
Alevin output incorrect matrix dimensions

### ISSUE THREAD

**Issue by Ryan-Zhu** [COMBINE-lab/salmon#380]
I ran salmon alevin 0.14.0 with a custom transcriptome reference and the following options: --chromium --dumpFeatures --dumpMtx --whitelist mylist.txt

Everything ran through OK. However, when I tried to load the .mtx file with readMM() in R, I got the error: 
` Error: readMM(): column values 'j' are not in 1:nc `

When I tried to read directly the binary file into a matrix, there's a warning:
```
counts <- readBin(quants_mat.gz, what = 'numeric', n = length(genes)*length(cells))
close.connection(quants_mat.gz)
Warning message:
In matrix(data = counts, nrow = length(cells), ncol = length(genes),  :
  data length [391335] is not a sub-multiple or multiple of the number of rows [4942]
```
It seems like there's issue with the dimensions
[out.zip](https://github.com/COMBINE-lab/salmon/files/3296080/out.zip)
 of the output matrix.

I've attached my log file as well as my output files (both .mtx and binary). Thank you.

**Comment by k3yavi** [comment]
Hi Ryan,
I was looking at the output matrix. It seems the count values are in scientific notations like e-7. Not sure how that happened and it's not happening at my end. I'll check what can be done. Thanks again for reporting this.

Regarding the binary file parsing, we recently optimize the output format for Alevin and have updated the python parser. We are still working on updating R parser, will update here once stable.

**Comment by Ryan-Zhu** [comment]
Hi Avi, thank you for your replay. Is below the python parser you are talking about? 
```
from vpolo.alevin import parser
alevin_df = parser.read_quants_bin("<PATH TO ALEVIN output folder>")
```
I'd like to export the data to .csv in python and then start from there. Thank you.

**Comment by k3yavi** [comment]
Hi Ryan,

Yes that's correct. You can try using the following R code for parsing too. Just heads up it might be a little slow, let me know if you have thoughts about vectorizing it. We generally recommend using tximport for importing Alevin output in R. However, while our R experts are working on updating tximport you can use the following piece of R code for parsing Alevin output from v0.14.0


[readAlevin.R.txt](https://github.com/COMBINE-lab/salmon/files/3298318/readAlevin.R.txt)
PS: ignore txt, had to add the extension to fool github.

**Comment by Ryan-Zhu** [comment]
Hi Avi,

I wish I had asked you sooner. The R parser works great. 

There is also a similar `readAlevin` function in Seurat v3 but it is not working with Alevin v0.14.0. It looks like it's the `# Version B specific support` section in your code that does the trick and I have no suggestion for vectorizing. I do suggest you to distribute the R parser more widely if possible because it will save users lots of frustrations . Several Alevin parsers I found online (e.g. Seurat, Alevin-tutorial paper) are all obsolete now. However, I understand this is work in progress and I hope my post here can also serve the purpose.

Thank you.

**Comment by k3yavi** [comment]
Thanks @Ryan-Zhu for your feedbacks and the suggestion.
I apologize for the trouble you had to face while working with the alevin output.
We will prepare better from the next release and try updating the external dependencies first before making an official release. 
Just wanted to give you the heads up that I have also updated the bug for the scientific notation in the `mtx` format. It's in the develop branch of alevin, if you have time please let me know if it works for you.

Thanks again.

**Comment by alexvpickering** [comment]
The issue with reading the .mtx file does not appear to be fixed in develop. I still get:

`Error: readMM(): column values 'j' are not in 1:nc` with `Matrix::readMM(file)`

I think it is related to the column indexing in the .mtx file. Specifically, values should be in `1:nc` where `nc` is in the first row second column after the `%%MatrixMarket  matrix  coordinate      real    general` line. There is a value of 0 in my .mtx file so I am guessing the issue is 0 indexing instead of 1 indexing.

**Comment by k3yavi** [comment]
Thanks again @alexvpickering, just pushed a fix.

**Comment by alexvpickering** [comment]
Np - thanks @k3yavi. Here is also a slightly faster (though still painfully slow) sparse reader for `quants_mat.gz`:

```R
readAlevin_sparse <- function(files) {
  dir <- sub("/alevin$","",dirname(files))
  barcode.file <- file.path(dir, "alevin/quants_mat_rows.txt")
  gene.file <- file.path(dir, "alevin/quants_mat_cols.txt")
  matrix.file <- file.path(dir, "alevin/quants_mat.gz")
  for (f in c(barcode.file, gene.file, matrix.file)) {
    if (!file.exists(f)) {
      stop("expecting 'files' to point to 'quants_mat.gz' file in a directory 'alevin'
  also containing 'quants_mat_rows.txt' and 'quant_mat_cols.txt'.
  please re-run alevin preserving output structure")
    }
  }
  cell.names <- readLines(barcode.file)
  gene.names <- readLines(gene.file)
  num.cells <- length(cell.names)
  num.genes <- length(gene.names)

  # mat <- matrix(nrow=num.genes, ncol=num.cells, dimnames=list(gene.names, cell.names))
  values <- rowinds <- colinds <- c()
  con <- gzcon(file(matrix.file, "rb"))

  get_binary <- function(id) { as.integer(head(intToBits(id), 8)) }

  pb <- utils::txtProgressBar(style = 3, max = num.cells)


  # Version B specific support
  num.bitvecs <- ceiling(num.genes/8)
  for (j in seq_len(num.cells)) {
    # read the bit vector
    bit.vec <- readBin(con, integer(), size=1, signed=FALSE, endian = "little", n=num.bitvecs)

    #iterating the bit vector
    gene.flags <- unlist(lapply(bit.vec, get_binary))
    num.exp.genes <- sum(gene.flags)

    # read in the expression of expressed genes
    counts <- readBin(con, double(), size=4, endian = "little", n=num.exp.genes)
    is.exp <- which(gene.flags == 1)

    values <- c(values, counts)
    rowinds <- c(rowinds, is.exp)
    colinds <- c(colinds, rep(j, length(is.exp)))

    utils::setTxtProgressBar(pb, j)
  }
  in.genes <- rowinds <= num.genes

  mat <- Matrix::sparseMatrix(i = rowinds[in.genes],
                              j = colinds[in.genes],
                              x = values[in.genes],
                              dims = c(num.genes, num.cells),
                              dimnames = list(gene.names, cell.names))


  close(con)
  mat
}

```

in comparison to version above:

```R
system.time(orig_mat <- readAlevin(files))
   user  system elapsed 
125.666   0.138 125.787 

system.time(sparse_mat <- readAlevin_sparse(files))
   user  system elapsed 
 71.821   0.164  71.905 

all.equal(as.matrix(sparse_mat), orig_mat)
[1] TRUE
```

**Comment by k3yavi** [comment]
Hi guys,
Thanks @alexvpickering for the faster version !
There has been a lot of language switches recently (C++, Rust, python and R), and I apologize there is a minor bug in the above code, thanks to @mikelove for pointing it.
```
 get_binary <- function(id) { as.integer(head(intToBits(id), 8)) }
```

should be updated to
```
 get_binary <- function(id) {  rev( as.integer(head(intToBits(id), 8)) ) }
```

The R functions returns least significant bit first while I was in the pythonic/C++ mode and assumed their native format of most significant bit first. Python parser was fine but please consider updating the above lines in the parser. @mikelove is double checking the code and will update the tximport package to make it more streamlined. I will keep you guys updated.

re: mtx format, yea I think we designed the sparse format mainly with the motivation of saving space both in memory and disk as our per cell level inferential data grows linearly with the number of bootstraps used.

**Comment by alexvpickering** [comment]
> Thanks again @alexvpickering, just pushed a fix.

Confirm that .mtx.gz file is loading properly now. Thanks again!

**Comment by k3yavi** [comment]
Awesome !! Appreciate the confirmation .

**Comment by k3yavi** [comment]
Hey guys, just wanna give heads up, we just released `0.14.1` for various bugfixes. Please consider updating salmon to the latest.
